### DISCUSSION TITLE
Implementation Tutorial for Neural Radiosity [Hadadan et al. 2021] in Mitsuba 3

### DISCUSSION THREAD

**Discussion by ciy405x** [mitsuba-renderer/mitsuba3#723] [Show and tell]
Hello folks,

For a recent project, I needed to implement a novel rendering technique called [Neural Radiosity (SIGGRAPH 2021)](https://dl.acm.org/doi/abs/10.1145/3478513.3480569) using Mitsuba 3. 

https://github.com/krafton-ai/neural-radiosity-tutorial-mitsuba3/blob/main/neural_radiosity.ipynb

In the Python notebook above, I provide a step-by-step guide to implementing NR, including sampling surface interactions uniformly on a mesh, integrating neural nets and Mitsuba 3, training with gradient descent, and writing code for various visualizations. These tasks are relevant not only to NR but also to various Mitsuba-PyTorch fusion projects. I believe other users will find it helpful as I utilized various methods from Mitsuba and Dr. Jit. 

While Mitsuba is a fantastic rendering library, there are few open-source codes or tutorials available, which made implementation challenging. Fortunately, with the help of the fantastic Discussions section of this repo, as well as the dedication of programmers and researchers at EPFL, I was able to complete the code.

Recently, some Show-and-tell posts have been published by its users, and I would like to join in and help others. :)
I hope to see more Mitsuba 3 open-source projects in the future as well.

**Comment by DoeringChristian** [comment]
Hey, thank you so much for the tutorial! It already helped me a lot. 
One note, you could set the shape pointer of the SurfaceInteraction i.e. `si.shape = shape` in the `sample_si` function.
This would allow you to query the BSDF later using the `si.bsdf()` function.

**Reply by ciy405x** [reply]
Oops! Thanks for notifying me of it. I'll change it soon!

**Reply by saeedhd96** [reply]
Thanks, this is very nice, please also note that the encoding is not getting 'otype' and therefore the encoding is by default 'OneBlob' and not 'HashGrid'.

**Reply by ciy405x** [reply]
OMG; that was a serious mistake while copy-pasting my original code. Thanks a lot!! @saeedhd96

**Reply by DoeringChristian** [reply]
Another "issue" I have come across as I was trying to render scenes with the principled BSDF  is that this BSDF cannot be rendered correctly. This is because in the `first_non_specular_or_null_si` function you are testing weather the sampled BSDF contains the `Smooth` flag. 
The principled BSDF samples from multiple lobes of which some are Smooth and others are Specular/Glossy.
To solve this I would suggest to sample the BSDF first and test the type of the sample. 
This way, the network only learns the diffuse part and when rendering we can randomly sample the lobe to use. I don't know what is used in the original implementation maybe @saeedhd96 could elaborate on this.
This is my proposed fix for the `first_non_specular_or_null_si` function:
```python
    def first_non_specular_or_null_si(
        self, scene: mi.Scene, si: mi.SurfaceInteraction3f, sampler: mi.Sampler
    ):
        """Find the first non-specular or null surface interaction.

        Args:
            scene (mi.Scene): Scene object.
            si (mi.SurfaceInteraction3f): Surface interaction.
            bsdf (mi.BSDF): BSDF object.
            sampler (mi.Sampler): Sampler object.

        Returns:
            tuple: A tuple containing four values:
                - si (mi.SurfaceInteraction3f): First non-specular or null surface interaction.
                - β (mi.Spectrum): The product of the weights of all previous BSDFs.
                - null_face (bool): A boolean mask indicating whether the surface is a null face or not.
        """
        # Instead of `bsdf.flags()`, based on `bsdf_sample.sampled_type`.
        with dr.suspend_grad():
            bsdf_ctx = mi.BSDFContext()

            depth = mi.UInt32(0)
            β = mi.Spectrum(1)
            # prev_si = dr.zeros(mi.SurfaceInteraction3f)
            # prev_bsdf_pdf = mi.Float(1.0)
            # prev_bsdf_delta = mi.Bool(True)

            null_face = mi.Bool(True)
            active = mi.Bool(True)

            loop = mi.Loop(
                name="first_non_specular_or_null_si",
                state=lambda: (sampler, depth, β, active, null_face, si),
            )
            loop.set_max_iterations(6)

            while loop(active):
                # for i in range(6):
                # loop invariant: si is located at non-null and Delta surface
                # if si is located at null or Smooth surface, end loop

                bsdf: mi.BSDF = si.bsdf()

                bsdf_sample, bsdf_weight = bsdf.sample(
                    bsdf_ctx, si, sampler.next_1d(), sampler.next_2d(), active
                )
                null_face &= ~mi.has_flag(
                    bsdf_sample.sampled_type, mi.BSDFFlags.BackSide
                ) & (si.wi.z < 0)
                active &= si.is_valid() & ~null_face
                active &= mi.has_flag(bsdf_sample.sampled_type, mi.BSDFFlags.Glossy)

                ray = si.spawn_ray(si.to_world(bsdf_sample.wo))

                si[active] = scene.ray_intersect(
                    ray,
                    ray_flags=mi.RayFlags.All,
                    coherent=dr.eq(depth, 0),
                    active=active,
                )

                β[active] *= bsdf_weight
                depth[active] += 1

        # return si at the first non-specular bounce or null face
        return si, β, null_face
```
Unfortunately I then cannot get other materials to work, using the `Smooth` flag did not work (I tested against the `Glossy` flag). 
![Figure_1](https://github.com/mitsuba-renderer/mitsuba3/assets/23581448/83a79e43-8fb1-4e30-b9e6-b59d47463b09)
This is the output that I rendered with 2spp (due to memory limitations I could not render it with more samples which would necessary, since now samples are split between diffuse and specular samples).

**Reply by ciy405x** [reply]
@DoeringChristian, could you provide me with the scene you've tested in the thread??

**Reply by DoeringChristian** [reply]
Of course. 
[cornell-box.zip](https://github.com/mitsuba-renderer/mitsuba3/files/11495254/cornell-box.zip)
I also modified your implementation a little further [here](https://gist.github.com/DoeringChristian/e396d30efeb4a0730c9a5b6888617402) you can find the one I used to render the image.

**Reply by ciy405x** [reply]
Kkk thx thx! Seems you are enjoying the neural radiosity. ;) Also, I made a few more changes in the Jupyter Notebook repository. Feel free to check it out.

**Reply by DoeringChristian** [reply]
Hi,
I had a problem where in some scenes evaluation takes a very long time.
It turns out that `loop.set_max_iterations` does not do anything in Mega-kernel mode. My fix was to add a test against the depth to the `active` variable.
This is especially a problem in scenes that have mirrors with concave geometry or glass like materials.
```python
active &= si.is_valid() & ~null_face & (depth < 6)
```

**Reply by ciy405x** [reply]
@DoeringChristian  I faced a similar problem before, but then I wonder what is the purpose of `loop.set.maxt_iterations`. Does it work in another mode?

**Reply by DoeringChristian** [reply]
@ciy405x 
As I understand it, in wavefront-style modes it disables the reduction of the condition at the end and only compares against a internal index (see [here](https://github.com/mitsuba-renderer/drjit/blob/bfab9ac11736b69fb8b5c0f7d9614fbfd82d627a/include/drjit/loop.h#L387)).
Unfortunately, the only documentation I could find was [here](https://github.com/mitsuba-renderer/drjit/blob/31a83a999a765eda671107761290eda615f8cb81/include/drjit/loop.h#L183).
```
/**
* \brief Set a fixed iteration count
*
* Only applies to wavefront-style loops. When the total number of loop
* iterations is known, call this function to avoid a costly horizontal
* reduction after each iteration.
*/
```

**Reply by DoeringChristian** [reply]
Btw. do you know why rendering caustics with this approach is not working as described in the paper?
![Caustics](https://github.com/mitsuba-renderer/mitsuba3/assets/23581448/fa66fdd7-61ac-4b13-bc66-85088cfd4466)
Here I rendered a simple scene that should produce caustics (3 hidden layers, width of 256, using Hash Grid encoding, and all other parameters the same as in your tutorial).
In the top left is the image rendered using Neural Radiosity, in the top right is an image rendered using the `ptracer` integrator and in the bottom left image you can see the reference rendered with the `path` integrator.
I optimized over 10.000 iterations (in the bottom left image is the loss).
Similarly, for caustics through glass (again with 10.000 training steps):
![ssd](https://github.com/mitsuba-renderer/mitsuba3/assets/23581448/07e62304-3fae-49fd-90c6-8a847b8cd345)

Thanks for the Help.

**Reply by ciy405x** [reply]
@DoeringChristian Wait... I guess my RHS implementation may be wrong. I'm checking this right now.

**Reply by ciy405x** [reply]
![image](https://github.com/mitsuba-renderer/mitsuba3/assets/132874228/ad862f78-e851-4931-a59d-c09701ddaaab)
Bug fixed. Thanks very much for your bug finding ;) I will update the main repository in minutes.

TL;DR:
Change a part of `render_rhs` function:
```python
si, β2, null_face = first_non_specular_or_null_si(scene, si, sampler)
ds = mi.DirectionSample3f(scene, si=si, ref=prev_si)
mis = mis_weight(
    prev_bsdf_pdf,
    scene.pdf_emitter_direction(prev_si, ds, ~prev_bsdf_delta),
)
```
as follow:
```python
ds = mi.DirectionSample3f(scene, si=si, ref=prev_si)
mis = mis_weight(
    prev_bsdf_pdf,
    scene.pdf_emitter_direction(prev_si, ds, ~prev_bsdf_delta),
)
si, β2, null_face = first_non_specular_or_null_si(scene, si, sampler)
```.

**Reply by DoeringChristian** [reply]
Oh, that makes sense thanks for the quick fix.

**Reply by DoeringChristian** [reply]
Hi,
I tested your implementation further and everything seems to work. However, as far as I can tell you are only constructing M/2 RHS samples per LHS sample i.e. `_r_sampler.seed(step, batch_size * M // 2)`. Is there a reason for this?

**Reply by ciy405x** [reply]
@DoeringChristian 
It depends on the interpretation of the Neural Radiosity paper. In Section 4.7 of the paper, it states that BSDF and emitter sampling are used to sample the direction of the RHS (Right-Hand Side), and a total of M samples are taken. Therefore, BSDF and emitter sampling are performed M/2 times each to construct M direction vectors. However, even when using the next event estimation in MC (Monte Carlo) path tracing, it is common to denote it as 1 spp (sample per pixel). Similarly, it may still be referred to as M spp if BSDF and emitter sampling are performed M times each. If you prefer the latter interpretation, you can double the value of M. Of course, it would be even better if the author @saeedhd96  could provide a direct explanation.

**Reply by saeedhd96** [reply]
Hi, we have released the neural radiosity code: https://github.com/saeedhd96/neural-radiosity
